### DISCUSSION TITLE
Twinbuilder - PyAEDT Queries

### DISCUSSION THREAD

**Discussion by shabham22** [ansys/pyaedt#5445] [Q&A]
Hi,

### 1. As shown in below image, I want to enable Both - NameAndValue property for the component using PyAEDT.

Can anyone tell how can I do this?

![image](https://github.com/user-attachments/assets/5ffa00f2-43df-4aae-a179-2524ec85f111)

### 2. Is it possible to fetch the Plot values or coordinates after generating the report with the help of PyAEDT. How?

Like from below example: _200kHz: 7328.36ohms and 700kHz: 26835.46ohms_
![image](https://github.com/user-attachments/assets/1495c508-55c8-4a9b-aad0-a58b8171dfab)

### 3. Is there any function or Module in PyAEDT to use the feature of Sheet Scan of any provided plot, to automatically export csv of coordinates.

![image](https://github.com/user-attachments/assets/1d3b1e58-561e-4c03-b6f1-91637fee5ac4)

**Comment by ghost** [comment]
Hello,

As for *question 1*, you might want to use **the modeler** attribute in the TwinBuilder application. Diving deep into it, you might find a function called **ChangeProperty** which takes an array as parameter:

```
app.modeler.oeditor.ChangeProperty(
	["NAME:AllTabs",
		["NAME:PropDisplayPropTab",
			["NAME:PropServers",
			resistor.composed_name],
			["NAME:NewProps",
				["NAME:R",
				"Format:=", "NameAndValue",
				"Location:=", "Bottom"]
			]
		]
	])
```

I would answer *questions 3 and 4* in one go by saying that once you define a setup for your model, then create a solution data object, you should be able to plot, generate a report which you can export into csv, image, or any other file.
The other, more general way around ids to manipulate the solution data object. You may look inside its data structure and see that all the piece of data required for a proper csv generation is inside: the axes with their unit 

x: ```solution_data.intrinsics["Time"]``` and the label: ```next(iter(x.units_sweeps.keys()))```, the unit: ```x.units_sweeps[next(iter(x.units_sweeps.keys()))]```
as for the y axis, ```solution_data.units_data[expression]``` for the units of any expression you are trying to look into, and the values would be hiding somewhere in ```list(x.full_matrix_real_imag[0][exp].values())```

please find an example of what the code would look like (as far as I'm concerned, exporting files directly from the solution data seemed to work with IronPython, but it does not seem to work very well when adapting the code to CPython.

Lastly I will invite you to look into a very useful tool in AEDT when you want to look into a certain feature in AEDT and want to automize it with Python: it is called _Record Script_ and can be foundin the icon bar Tab 'Automation"




```
import os

import csv
from ansys.aedt.core import TwinBuilder
import time

AEDT_VERSION = "2024.2"
NUM_CORES = 4
NG_MODE = True

base_path = 'C:\Atelier_Personnel\PythonCode\AEDT\pythonProject\\res'
project_name = os.path.join(base_path, "rc_circuit.aedt")
app = TwinBuilder(
    project=project_name,
    version=AEDT_VERSION,
    non_graphical=NG_MODE,
    new_desktop=True,
)

app.modeler.schematic_units = 'mil'
schematic = app.modeler.schematic
source = schematic.create_voltage_source("E1", "EPULSE", 10, 10, [0, 0])
resistor = schematic.create_resistor("R1", 10000, [1000, 1000], 90)
capacitor = schematic.create_capacitor("C1", 1e-6, [2000, 0])
gnd = app.modeler.components.create_gnd([0, -1000])

source.pins[1].connect_to_component(resistor.pins[0], use_wire=True)
resistor.pins[1].connect_to_component(capacitor.pins[0], use_wire=True)
capacitor.pins[1].connect_to_component(source.pins[0], use_wire=True)
source.pins[0].connect_to_component(gnd.pins[0], use_wire=True)

app.modeler.oeditor.ChangeProperty(
	["NAME:AllTabs",
		[
			"NAME:PropDisplayPropTab",
			[
				"NAME:PropServers",
				resistor.composed_name
			],
			[
				"NAME:NewProps",
				[
					"NAME:R",
					"Format:="		, "NameAndValue",
					"Location:="		, "Bottom"
				]
			]
		]
	])
app.set_end_time("300ms")
app.analyze_setup("TR")

E_Value = "E1.V"
C_Value = "C1.V"
expressions = [E_Value, C_Value]

post = app.post
x = post.get_solution_data(expressions, "TR", "Time")
x.plot([E_Value, C_Value], x_label="Time", y_label="Capacitor Voltage vs Input Pulse")
report = post.create_report(expressions=[E_Value, C_Value], setup_sweep_name="TR", domain="Time", plot_name="Rectangular")
# post.export_report_to_csv(project_dir=os.path.join(base_path, 'rc_circuit'), plot_name="Capacitor Voltage vs Input Pulse")
time_scale = x.intrinsics["Time"]

result = []
import matplotlib.pyplot as plt
for exp in expressions:
	plt.plot(time_scale, x.data_real(exp))
	result.append(list(x.full_matrix_real_imag[0][exp].values()))
rows = []
for i in range(len(time_scale)):
	rows.append([time_scale[i]] + list([res[i] for res in result]))
with open(os.path.join(base_path, 'rc_circuit.csv'), mode="w", newline="") as file:
	writer = csv.writer(file)
	# Write the header
	time_title = next(iter(x.units_sweeps.keys()))
	header = [F"\"{time_title} [{x.units_sweeps[time_title]}]\""] + [f"\"{exp} [{x.units_data[exp]}]\"" for exp in expressions]
	writer.writerow(header)
	# Write the rows
	writer.writerows(rows)

plt.xlabel("Time")
plt.ylabel(E_Value + ", " + C_Value)
plt.title("Capacitor Voltage vs Input Pulse")
plt.savefig(os.path.join(base_path, 'rc_circuit.jpg'))
app.save_project()
app.release_desktop()
# Wait 3 seconds to allow AEDT to shut down before cleaning the temporary directory.
time.sleep(3)
```
